yt_dlp/output/logging_output.py

```diff
@@ -0,0 +1,137 @@
+import inspect
+import logging
+import sys
+import traceback
+
+
+class deferred_class_attribute:
+    def __init__(self, func):
+        self.func = func
+
+    def __get__(self, _, owner):
+        result = self.func()
+        setattr(owner, self.func.__name__, result)
+        return result
+
+
+class _YDLIgnoringLogger(logging.Logger):
+    @deferred_class_attribute
+    def _internal_frames():
+        from . import logging as output_logging
+        from ..downloader import common as downloader
+        from ..extractor import common as extractor
+
+        return frozenset(map(inspect.getabsfile, [
+            _YDLIgnoringLogger,
+            output_logging,
+            logging,
+            downloader,
+            extractor,
+        ]))
+
+    @deferred_class_attribute
+    def _ydl_path():
+        from .. import YoutubeDL
+
+        return inspect.getabsfile(YoutubeDL)
+
+    @staticmethod
+    def skip_frames(condition, start_frame):
+        frame = start_frame
+        if frame is None:
+            return start_frame
+
+        depth = 0
+        while condition(frame, depth):
+            depth += 1
+            next_frame = frame.f_back
+            if next_frame is None:
+                return start_frame
+
+            frame = next_frame
+
+        return frame
+
+    @staticmethod
+    def _get_raising_frame():
+        error = sys.exc_info()[1]
+        if error is None:
+            return None
+
+        error_chain = [error]
+
+        while error is not None:
+            error_chain.append(error)
+            error = error.__cause__ or error.__context__
+
+        for error in reversed(error_chain):
+            tb = error.__traceback__
+            if tb is None:
+                continue
+            while tb.tb_next is not None:
+                tb = tb.tb_next
+
+            return tb.tb_frame
+
+        return None
+
+    @classmethod
+    def _get_logging_frame(cls, stacklevel):
+        frame = inspect.currentframe()
+        if frame is None:
+            return None
+
+        def internal_frame(frame, _):
+            return inspect.getabsfile(frame) in cls._internal_frames
+
+        frame = cls.skip_frames(internal_frame, frame)
+
+        # We need to skip additional frames since the
+        # logging call could be delegated through YDL.
+        # We look for the highest frame inside `YoutubeDL`,
+        # then until it is no longer internal
+        # so that `common` is skipped as well
+        def ydl_frame(frame, _):
+            return inspect.getabsfile(frame) != cls._ydl_path
+
+        above_ydl_frame = cls.skip_frames(ydl_frame, frame).f_back
+        # Assumption: YoutubeDL.py is never main
+        assert above_ydl_frame is not None
+        frame = cls.skip_frames(internal_frame, above_ydl_frame)
+
+        # Skip `stacklevel` amount of levels
+        frame = cls.skip_frames(lambda _, depth: depth == stacklevel, frame)
```

Comment:
Do we need to support `stacklevel > 1`? Otherwise this could just be `f_back` and we can simplify `skip_frames` a lot.